source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1027,6 +1028,194 @@ ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks&
+
+class ClusterCallbacks : public ClusterUpdateCallbacks {
+public:
+  ClusterCallbacks(ClusterAddedCb cb) : cb_(std::move(cb)) {}
+
+  void onClusterAddOrUpdate(ThreadLocalCluster& cluster) override { cb_(cluster); };
+
+  void onClusterRemoval(const std::string&) override {}
+
+private:
+  ClusterAddedCb cb_;
+};
+
+} // namespace
+
+ClusterManagerImpl::ClusterDiscoveryManager::ClusterDiscoveryManager(
+    ThreadLocalClusterManagerImpl& parent)
+    : parent_(parent) {}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::ensureLifecycleCallbacksAreInstalled() {
+  if (callbacks_handle_) {
+    return;
+  }
+  auto cb = ClusterAddedCb([this](ThreadLocalCluster& cluster) {
+    processClusterName(cluster.info()->name(), ClusterDiscoveryStatus::Available);
+  });
+  callbacks_ = std::make_unique<ClusterCallbacks>(cb);
+  callbacks_handle_ = parent_.parent_.addThreadLocalClusterUpdateCallbacks(*callbacks_);
+}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::processClusterName(
+    const std::string& name, ClusterDiscoveryStatus cluster_status) {
+  // Extracting the list of callbacks from the map makes resetting the
+  // handle inside the callback safe, because handle would try to find
+  // the list of callbacks in the map and would find nothing, instead
+  // of removing an item from the list while iterating the list.
+  auto map_node_handle = pending_clusters_.extract(name);
+  if (map_node_handle.empty()) {
+    return;
+  }
+  for (const auto& weak_callback : map_node_handle.mapped()) {
+    auto callback = weak_callback.lock();
+    if (callback != nullptr) {
+      (*callback)(cluster_status);
+    }
+  }
+  maybePostResetCallbacks();
+}
+
+ClusterManagerImpl::ClusterDiscoveryManager::Pair
+ClusterManagerImpl::ClusterDiscoveryManager::addCallback(
+    const std::string& name, const ClusterDiscoveryCallbackSharedPtr& callback) {
+  ensureLifecycleCallbacksAreInstalled();
+
+  auto& callbacks_list = pending_clusters_[name];
+  auto it = callbacks_list.emplace(callbacks_list.end(), callback);
+  auto handle = std::make_unique<ClusterDiscoveryCallbackHandleImpl>(*this, name, it);
+  auto discovery_in_progress = (callbacks_list.size() > 1);
+  return {std::move(handle), discovery_in_progress};
+}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::erase(const std::string& name,
+                                                        CallbackListIterator it) {
+  const bool drop_list = eraseFromList(name, it);
+  if (drop_list) {
+    pending_clusters_.erase(name);
+  }
+  maybePostResetCallbacks();
+}
+
+bool ClusterManagerImpl::ClusterDiscoveryManager::eraseFromList(const std::string& name,
+                                                                CallbackListIterator it) {
+  auto map_it = pending_clusters_.find(name);
+  if (map_it == pending_clusters_.end()) {
+    return false;
+  }
+  auto& list = map_it->second;
+  list.erase(it);
+  return list.empty();
+}
```

Comment:
Can you take a look at `RaiiListElement` and `RaiiMapOfListElement` in cleanup.h and see if they can be of help here? This is very similar to logic as we already have and it would be easier to rely on those helpers if possible.

Reply:
I'll check if I can use the `RaiiListElement` - `RaiiMapOfListElement` assumes that it is the only way to destroy the item, so it asserts that the key in map actually exists. Which usually is not true in our case.

Reply:
`RaiiListElement` would also be awkward to use, as I ended up using `cancel` and `erase` directly instead of relying on its destructor. The code was more or less like:
```c++
if (!map.contains(cluster_name)) {
  raii.cancel()
  return;
}
raii.erase();
list = map[cluster_name]
if (list.empty()) {
  map.erase(cluster_name);
}
```

So it does not buy me much here in the end.

Reply:
OK, thanks for checking.